packages/medusa/src/services/payment-provider.ts

```diff
@@ -696,4 +747,114 @@ export default class PaymentProviderService extends TransactionBaseService {
 
     return refund
   }
+
+  /**
+   * Build the create session context for both legacy and new API
+   * @param cartOrData
+   * @protected
+   */
+  protected buildCreatePaymentContext(
+    cartOrData: Cart | PaymentSessionInput
+  ): Cart & PaymentContext {
+    const cart =
+      "object" in cartOrData && cartOrData.object === "cart"
+        ? cartOrData
+        : ((cartOrData as PaymentSessionInput).cart as Cart)
+
+    const context = {} as Cart & PaymentContext
+
+    // TODO: only to support legacy API. Once we are ready to break the API, the cartOrData will only support PaymentSessionInput
+    if ("object" in cartOrData && cartOrData.object === "cart") {
+      context.cart = {
+        context: cart.context,
+        shipping_address: cart.shipping_address,
+        id: cart.id,
+        email: cart.email,
+        shipping_methods: cart.shipping_methods,
+      }
+      context.amount = cart.total!
+      context.currency_code = cart.region?.currency_code
+      context.collected_data = cart.customer?.metadata ?? {}
+      Object.assign(context, cart)
+    } else {
+      const data = cartOrData as PaymentSessionInput
+      context.cart = data.cart
```

Comment:
thought: I would suggest that we never pass the entire cart to the payment provider plugin, to avoid provider plugins heavily relying on that data structure. What do you think of that?